R/importance.R

```diff
@@ -36,131 +39,243 @@
+    }
 
-  # X-length list of importance scores
-  res_list <- lapply(X, function(x) {
+    # check that all covariates in the groups are also in the task's covariates
+    if (!all(unlist(covariate_groups) %in% X)) {
+      stop("Groups contain covariates that are not in the task's covariates.")
+    }
+
+    # check if all covariates in task are in groups, and if not add them
+    if (!all(X %in% unlist(covariate_groups))) {
+      missingX <- as.list(X[which(!X %in% unlist(covariate_groups))])
+      names(missingX) <- X[which(!X %in% unlist(covariate_groups))]
+      covariate_groups <- c(covariate_groups, missingX)
+    }
+    # check that groups with more than one covariate are named, and add name to
+    # unnamed groups but only if unnamed group is a single covariate
+    if (any(is.null(names(covariate_groups))) |
+      any(names(covariate_groups) == "") |
+      any(is.na(names(covariate_groups)))) {
+      no_name <- unique(which(
+        is.null(names(covariate_groups)) | names(covariate_groups) == "" |
+          is.na(names(covariate_groups))
+      ))
+      if (any(sapply(covariate_groups[unique(no_name)], length)) != 1) {
+        stop("Covariate groups with more than one covariate must be named.")
+      } else if (all(sapply(covariate_groups[unique(no_name)], length)) == 1) {
+        names(covariate_groups[no_name]) <- unlist(covariate_groups[no_name])
+      }
+    }
+    X <- covariate_groups # for convenience in remaining parts of function
+  } else {
+    names(X) <- X
+  }
+
+  ########################## risk under observed data ##########################
+  original_pred <- fit$predict_fold(task, fold_number = fold_number)
+  original_eval <- eval_fun(original_pred, Y)
+  if (!is.null(attr(original_eval, "loss")) && !attr(original_eval, "loss")) {
+    original_risk <- original_eval
+  } else {
+    original_losses <- original_eval
+    original_risk <- mean(original_losses)
+  }
+
+  ######################## list of importance results ##########################
+  res_list <- lapply(seq_along(X), function(i) {
+    # get the relevant group/covariate
+    if (!is.null(covariate_groups)) {
+      x <- X[[i]]
+    } else {
+      x <- X[i]
+    }
+    # get the risk when the covariate/group is permuted/removed
     if (type == "permute") {
-      # permute covariate x, and give it the same name as the original x
-      x_permuted <- data.table(sample(unlist(d[, x, with = FALSE]), nrow(d)))
-      names(x_permuted) <- x
-      # replace original x with permuted x, and update task with permuted x
-      x_permuted_name <- task$add_columns(x_permuted)
-      task_x_permuted <- task$next_in_chain(column_names = x_permuted_name)
-      # obtain predictions & risk on the new task with permuted x
-      x_permuted_pred <- fit$predict_fold(task_x_permuted, fold_number)
-      no_x_risk <- mean(loss(x_permuted_pred, Y))
+      seeds <- sample(11:2^15, B)
+      no_x_risk <- mean(unlist(parallel::mclapply(seeds, function(b) {
```

Comment:
change `parallel::mclapply` to `future.apply::future_lapply`. would also recommend chaining the output for readability, i.e., instead of `mean(unlist(...))`, it's equivalent to just call `unlist(...)` then `mean(...)` with extra assignments in between.